### PR TITLE
Build qemu-vm without KVM. Require KVM only for execution of VM.

### PR DESCRIPTION
**Pull request by ts468**

### CODE REVIEW

**Comment by ts468** [comment]
@falsifian, you merged my PR #5995 a week ago. This PR has related changes, and so I wondered if you could maybe have a look at it as well?

File: nixos/modules/virtualisation/qemu-vm.nix [hunk 1]
```diff
@@ -126,8 +130,8 @@ let
             '';
           buildInputs = [ pkgs.utillinux ];
           QEMU_OPTS = if cfg.useEFIBoot
-                      then "-pflash $out/bios.bin -nographic -serial pty"
-                      else "-nographic -serial pty";
+                      then "-no-kvm -pflash $out/bios.bin -nographic -serial pty"
+                      else "-no-kvm -nographic -serial pty";
```
Comment (falsifian) [inline nixos/modules/virtualisation/qemu-vm.nix]: Why pass the -no-kvm flag here?

**Comment by falsifian** [comment]
I'm confused --- won't qemu_kvm still be built if you try to use this module, since it's used by the startVM script? What does this commit let you build that won't cause qemu_kvm to be built?

Also, it looks like this change would cause two copies of qemu to exist in the store.

I don't think I have the full context here... it would probably be better to page someone who's worked on this file. (Though I'm personally curious to know the context, i.e. why you want the change.)

**Comment by ts468** [comment]
@falsifian, thanks for reviewing the PR. The problem that I tried to solve is that currently you have to have qemu-kvm available to _build_ the VM. By taking the non-kvm version of qemu for building the VM, qemu-kvm is only needed for actually running the VM.

**Comment by falsifian** [comment]
But it looks like KVM is still required, even with this PR.  If I edit ```qemu/default.nix``` to assert ```!x86only``` and then run

    nix-build -A vm nixos

I see an assertion failure. Is there some other thing this allows you to build without KVM?

**Comment by falsifian** [comment]
I think this happens because the build script produced refers to a binary under qemu_kvm.

Is KVM really big or slow to build?

**Comment by falsifian** [comment]
I guess by KVM I mean the KVM-enabled version of qemu.

**Comment by ts468** [comment]
And I meant KVM enabled system/host. So the idea is to be able to build the VM in a non-KVM-enabled system. I ran into that problem when I was working on a Xen-enabled system where the KVM related modules were not loaded or not available.

**Comment by ts468** [comment]
Thanks for reviewing the PR!

**Comment by falsifian** [comment]
Ah, it makes sense now! So the qemu binary won't even run on non-KVM systems, even if you're just creating a disk image? (Did you try adding -no-kvm to the disk-image-creating command?) Thanks for fixing this.

I still have a comment asking why you're adding -no-kvm to QEMU_OPTS, which looks like it's only used when the VM is run. (Indeed, it might be desirable to run the VM on non-KVM systems, but that seems like a separate issue...? From what you say, it sounds like the command won't run anyway on non-KVM systems because it's from pkgs.qemu_kvm.)

Since I've never touched this stuff... @bluescreen303 or @edolstra, are you happy with the VM module depending on both qemu and qemu_kvm, with the benefit that it can be built on non-KVM systems? If so I guess I will merge it.

**Comment by falsifian** [comment]
In addition to the above, could you add a comment explaining that you use non-KVM qemu to build the disk image so that the expression can evaluate on systems without KVM? Otherwise some helpful person might come along later and change it back to qemu_kvm in order to make the closure smaller :-)

**Comment by ts468** [comment]
@falsifian, thanks for your feedback and efforts for getting this PR merged, and am very sorry to say that I probably made a lot of noise for nothing. When I was trying to work on your comments, my previous problems seemed to be not existent anymore---everything works smoothly without further changes. I don't know what led to my problems in the first place, but for now I'll close the PR. Sorry about that!

**Comment by falsifian** [comment]
No problem, @ts468. I think it's better to rais the issue than to ignore problems you are encountering, so thanks for investigating :-)

**Pull request closed without merging**